python/cuml/common/array.py

```diff
@@ -0,0 +1,259 @@
+    strides : tuple of ints
+        Strides of the data
+    __cuda_array_interface__ : dictionary
+        __cuda_array_interface__ to interop with other libraries.
+
+    Object Methods
+    --------------
+
+    to_output : Convert the array to the appropriate output format.
+
+    Class Methods
+    -------------
+
+    Array.empty : Create an empty array, allocating a DeviceBuffer.
+    Array.zeros : Create an Array with allocated DeviceBuffer initialized with
+        zeros.
+
+    Notes
+    -----
+
+    cuml Array is not meant as an end-user array library. It is meant for
+    cuML/RAPIDS developer consumption. Therefore it contains the minimum
+    functionality. Its functionality is hidden by base.pyx to provide
+    automatic output format conversion so that the users see the important
+    attributes in whatever format they prefer.
+
+    Todo: support cuda streams in the constructor, particularly after
+
+
+
+    """
+
+    def __init__(self, data=None, owner=None, dtype=None, shape=None,
+                 order=None):
+        if data is None:
+            raise TypeError("To create an empty Array, use the class method" +
+                            " Array.empty()")
+        if isinstance(data, int):
+            if dtype is None or shape is None or order is None:
+                raise TypeError("Need to specify dtype, shape and order when" +
+                                " creating an Array from a pointer.")
+
+        ary_interface = False
+
+        if isinstance(data, DeviceBuffer) or isinstance(data, int):
+            size, shape = _get_size_from_shape(shape, dtype)
+            super(Array, self).__init__(data=data, owner=owner, size=size)
+            self.shape = shape
+            self.dtype = np.dtype(dtype)
+            self.order = order
+            self.strides = _order_to_strides(order, shape, dtype)
+
+        elif hasattr(data, "__array_interface__"):
+            ary_interface = data.__array_interface__
+
+        elif hasattr(data, "__cuda_array_interface__"):
+            ary_interface = data.__cuda_array_interface__
+
+        else:
+            raise TypeError("Unrecognized data type.")
+
+        if ary_interface:
+            super(Array, self).__init__(data=data, owner=owner)
+            self.shape = ary_interface['shape']
+            self.dtype = np.dtype(data.dtype)
+            if ary_interface['strides'] is None:
+                self.order = 'C'
+                self.strides = _order_to_strides(self.order, self.shape,
+                                                 self.dtype)
+            else:
+                self.strides = ary_interface['strides']
+                self.order = _strides_to_order(self.strides, data.dtype)
+
+    def __getitem__(self, slice):
+        return Array(data=cp.asarray(self).__getitem__(slice))
+
+    def __setitem__(self, slice, value):
+        cp.asarray(self).__setitem__(slice, value)
```

Comment:
Would add `__len__` as well. Dask looks for this kind of thing during serialization. Also can be handy for simplification of code. Normally NumPy-like libraries just return `self.shape[0]` for this. So pretty easy to implement.

Also Dask uses the `nbytes` property to estimate the size of an object if it is being transmitted. If this is always contiguous, it might be easy to leave this alone and just rely on the (soon to be added) `Buffer` implementation we have. If non-contiguous arrays are considered, this may be needed as the array size will be smaller.